## CS 156 HW 1
### Ankush Hommerich-Dutt

1. **[d]** 

Explanation: 

(1) This is not learning, since you **already know** the exact coin specifications from the mint. There is no learning that needs to be done by the machine, since it already knows by the statistical model on how to classify each coin. 

(2) This is supervised learning, since you are given a set of **labeled** data. It then uses this data to actual learn and create the specifications that should exist for each coin, so that future predictions can conform to these learned specifications.

(3) This is reinforcement learning, since the outcome of the move is not know immediately, but later the move selected is given a grade (whether the game was won or not), and the weights for the moves in the game are tweaked accordingly. After millions of games, this grade for each move helps the computer develop a strategy for tic-tac-toe.

2. **[a]** 

Explanation:
    
(1) This is not well suited for machine learning since there already exists a mathematical formula to classify primes, and it is not too difficult to evaluate.

(2) This is well suited because there is plenty of data available for whether credit card charges resulted in fraud, and the underlying mathematical formula for this is not known, and so a machine can learn from all the data to better detect whether a credit card charge is fraudulent.

(3) This is not well suited because there is nothing to be learned. There already exists formulas for this phenomenon (kinematics equations) which work very well for its purpose.

(4) This is well suited because there must exist a patten for this, this is a problem that does not have any good formulas existing for it already, and there can exist lots of data on the congestion of an intersection based on the traffic light cycle. Machine learning could get to an optimal solution by looking at large amounts of data much faster than a human ever could. 

3. **[d]** 

Bag 1 is the bag with only black balls, and bag 2 is the bag with one black and one white ball. Since you chose a black ball, you could have chosen either ball in bag 1, or the black ball in bag 2. This gives us 3 possibilities. Out of these 3 possibilities, 2 of them have the case where the second ball is also black (if the ball you initially picked was in bag 1).

4. **[b]**

The probability that a red ball isn't chosen (a green ball is chosen) is 1-0.55=0.45.

$(0.45)^{10} = 0.0003405 = 3.405 \times 10^{-4}$

5. **[c]** 

The probability that a sample has $\nu = 0$ is $3.405 \times 10^{-4}$. Therefore, the probability that a sample does not have $\nu = 0$ is $1 - 3.405 \times 10^{-4} = 0.9997$. If we run a thousand samples, the chance that none of the samples has $\nu = 0$ is $0.9997^{1000} = 0.7114$. Therefore, the chance that at least one of the samples has $\nu = 0$ is $1 - 0.7114 = 0.2886 = 0.289$.  

6. **[e]**

Based on the scoring system provided:

(a) (1 x 3) + (3 x 2) + (3 x 1) = 3 + 6 + 3 = 12

(b) (1 x 3) + (3 x 2) + (3 x 1) = 3 + 6 + 3 = 12

(c) (1 x 3) + (3 x 2) + (3 x 1) = 3 + 6 + 3 = 12

(d) (1 x 3) + (3 x 2) + (3 x 1) = 3 + 6 + 3 = 12

So therefore, all hypotheses for choices a-d have an equivalent score, and so e is the correct answer.



In [6]:
import random

class Point:
    def __init__(self, x, y):
        self.val = (x, y)
        
    def outcome(self, line):
        target = line.evaluate(self.val[0])
        if self.val[1] > target:
            return 1
        return -1
        
class Line:
    def __init__(self, point1, point2):
        self.slope = (point2.val[1] - point1.val[1]) / (point2.val[0] - point1.val[0])
        self.intercept = (self.slope * (-point1.val[0])) + point1.val[1]
        
    def evaluate(self, x):
        return (self.slope * x) + self.intercept
    
class Weight_vector:
    def __init__(self):
        self.vec = [0,0,0]
    
    def dot(self, point):
        sum = 0
        for i in range(3):
            if i == 0:
                sum += self.vec[i]
            else:    
                sum += self.vec[i] * point.val[i - 1]
        return sum   
    
    def update(self, y, x1, x2, x3):
        self.vec[0] += y*x1
        self.vec[1] += y*x2
        self.vec[2] += y*x3
        
def random_point():
    return Point(random.uniform(-1,1), random.uniform(-1,1))

def target_function():
    point1 = random_point()
    point2 = random_point()
    return Line(point1, point2)

def sign(x):
    if x < 0:
        return -1
    elif x == 0:
        return 0
    return 1

def iteration(outcomes, weight_vec, target):
    mis = [i[0] for i in outcomes if sign(weight_vec.dot(i[0])) != i[1]]
    if len(mis) == 0:
        return True
    bad_point = mis[random.randint(0,len(mis) - 1)]
    weight_vec.update(bad_point.outcome(target), 1, bad_point.val[0], bad_point.val[1])
    
    return False
    
def main(n):
    target = target_function()
    data = [random_point() for x in range(n)]
    outcomes =  list(map(lambda x: (x,x.outcome(target)), data))
    weight = Weight_vector()
    
    converge = False 
    num_iterations = 0
    while converge != True:
        converge = iteration(outcomes, weight, target)
        num_iterations += 1
        
    new_data = [random_point() for x in range(1000)]
    new_outcomes =  list(map(lambda x: (x,x.outcome(target)), new_data))
    mis = [i[0] for i in new_outcomes if sign(weight.dot(i[0])) != i[1]]
    prob = len(mis) / 1000
    
    return num_iterations, prob

random.seed(5)

test = [main(10) for i in range(1000)]

num_iterations = 0
prob = 0
for i in range(1000):
    num_iterations += test[i][0]
    prob += test[i][1]
    
print("for N = 10: \nThe number of iterations is:", num_iterations / 1000, 
      "and the probability of f(x) != g(x): ", prob / 1000)    



test2 = [main(100) for i in range(1000)]

num_iterations2 = 0
prob2 = 0
for i in range(1000):
    num_iterations2 += test2[i][0]
    prob2 += test2[i][1]
    
print("for N = 100: \nThe number of iterations is:", num_iterations2 / 1000, 
      "and the probability of f(x) != g(x): ", prob2 / 1000)  

for N = 10: 
The number of iterations is: 10.904 and the probability of f(x) != g(x):  0.11314700000000008
for N = 100: 
The number of iterations is: 110.125 and the probability of f(x) != g(x):  0.01432399999999998


7. **[b]** 

15.793 is closest to 15

8. **[c]** 

0.112 is closest to 0.1

9. **[b]** 

194.062 is closest to 100

10. **[b]** 

0.014 is closest to 0.01